### PR TITLE
Add 'non blocking' and 'with timeout' send and recieve over the NoC

### PR DESCRIPTION
**Pull request by ChadliaJerad**
This PR is to add nonblocking send and receive routines, as well as send and receive within a timeout.
They are useful for the predictability tests in the InterPRET platform.

### CODE REVIEW

**Comment by erlingrj** [comment]
Great suggestion, these additions look very useful but I would argue for the following API:
```
int noc_receive(uint32_t *recv, uint32_t timeout);
int noc_send(uint32_t addr, uint32_t data, uint32_t timeout);
```

The return value can be the status (success/timeout) and the data is written to the address pointed by `recv`. timeout=0 means try once, timeout=UINT32_MAX means blocking forever. I think we should only have a single receive and send function which takes a timeout parameter

File: programs/lib/include/flexpret_noc.h [hunk 1]
```diff
@@ -13,26 +14,83 @@
...
+    }
+    if (timeout == 0) {
+        if (NOC_DATA_AVAILABLE(NOC_CSR))
+            return NOC_DATA;
+        return 0;
+    }
+    uint32_t time = rdtime() + timeout;
+    while (rdtime() < time) {
+        if (NOC_DATA_AVAILABLE(NOC_CSR))
+            return NOC_DATA;
```
Comment (lsk567) [inline programs/lib/include/flexpret_noc.h:91]: This is open for discussion. But to improve predictability, it makes sense to me to make the `noc_receive()` constant-time by adding a `delay_until(time)` before returning. The current implementation enforces an upper bound, which could have timing variance if the function returns early. A `delay_until(time)` enforces the lower bound of execution time.

Comment (ChadliaJerad) [reply programs/lib/include/flexpret_noc.h:91]: I see your point, but I don't agree.
If we use `delay_until(timeout)`, then we may lose data if two cores send to the same core close enough in time (that is: both send happen at less than the timeout interval).
Better to catch up on the total execution time using `delay_until()` at the end of the task.
In addition, if the timeout will be equal to the trip duration for a message to arrive, then probably it is not worth it to put the hardware thread to sleep and then wake up.
Nevertheless, it is open for discussion, indeed.

File: programs/lib/include/flexpret_noc.h [hunk 2]
```diff
@@ -13,26 +14,83 @@
...
+ *
+ * Args:
+ *  * timeout: the timeout in ns
+ * Returns:
+ *  * uint32_t: the received data, 0 otherwise
+ **/
+static uint32_t noc_receive(uint32_t timeout) {
+    if (timeout == UINT32_MAX) {
+        while (!NOC_DATA_AVAILABLE(NOC_CSR));
+        return NOC_DATA;
```
Comment (erlingrj) [inline programs/lib/include/flexpret_noc.h:81]: It sould return SUCCESS and write NOC_DATA to a pointer

**Review by erlingrj — COMMENTED** [review]
Two very minor comments:

1. I would reorder the arguments: `int noc_receive(uint32_t *data, timeout_t timeout)`
2. Can we make our own custom return type for statuses? We need them all over the libraries. I propose for now:

```
typedef enum {
FP_SUCCESS=0,
FP_FAIL=1,
FP_TIMEOUT=2
FP_OUT_OF_MEMORY=3
} fp_ret_t
```

Then using our API would look like:

```
if (noc_receive(...) != FP_SUCCESS) {
   ...
}
```


Also I would propose to add some definitions

```
typedef timeout_t uint32_t
#define TIMEOUT_FOREVER UINT32_MAX
#define TIMEOUT_NEVER 0
```

**Comment by ChadliaJerad** [comment]
> Two very minor comments:
> 
> 1. I would reorder the arguments: `int noc_receive(uint32_t *data, timeout_t timeout)`
Makes sense.
> 2. Can we make our own custom return type for statuses? We need them all over the libraries. I propose for now:
> 
> ```
> typedef enum {
> FP_SUCCESS=0,
> FP_FAIL=1,
> FP_TIMEOUT=2
> FP_OUT_OF_MEMORY=3
> } fp_ret_t
> ```
> 
It is unclear when to have what.
`FP_TIMEOUT` and `FP_OUT_OF_MEMORY` are particular cases of `FP_FAIL`. 
When there is a timeout send or receive, the operation may fail because of the timeout, or because of the memory.
Do you mean this (in almost pseudo-code) :
```
static uint32_t noc_receive(uint32_t timeout, uint32_t* data) {
    if (timeout == UINT32_MAX) {
        while (!NOC_DATA_AVAILABLE(NOC_CSR));
        *data = NOC_DATA;
        return FP_SUCCESS;
    }
    if (timeout == 0) {
        if (NOC_DATA_AVAILABLE(NOC_CSR))
            *data = NOC_DATA;
            return FP_SUCCESS;
        return FP_OUT_OF_MEMORY;
    }
    uint32_t time = rdtime() + timeout;
    while (rdtime() < time) {
        if (NOC_DATA_AVAILABLE(NOC_CSR))
            *data = NOC_DATA;
            return FP_SUCCESS;
    }
    return FP_TIMEOUT; // Here it is both, the timeout and the memory!
}
```
If yes, when will `FP_FAIL` be returned?


> Also I would propose to add some definitions
> 
> ```
> typedef timeout_t uint32_t
> #define TIMEOUT_FOREVER UINT32_MAX
> #define TIMEOUT_NEVER 0
> ```
Yes, why not.

**Comment by erlingrj** [comment]
A little unclear from me. `FP_OUT_OF_MEMORY` is an unrelated error code for other libraries. `FP_FAIL` will be the generic fail return code. In noc_receive FP_FAIL should be returned of timout=NEVER and there is no data/FIFO is full

**Comment by ChadliaJerad** [comment]
> A little unclear from me. `FP_OUT_OF_MEMORY` is an unrelated error code for other libraries. `FP_FAIL` will be the generic fail return code. 

Got the answer to the question.

File: programs/lib/include/flexpret_types.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// #include <stdint.h>
```
Comment (erlingrj) [inline programs/lib/include/flexpret_types.h:1]:
```suggestion
#include <stdint.h>
```

Comment (erlingrj) [reply programs/lib/include/flexpret_types.h]: You do need stdint.h

Comment (ChadliaJerad) [reply programs/lib/include/flexpret_types.h:1]: Indeed!

File: programs/lib/include/flexpret_types.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+typedef uint32_t timeout_t;
+
+// Timeout definitions
+#define TIMEOUT_FOREVER UINT32_MAX
+#define TIMEOUT_NEVER 0
+
+// Return types
+typedef enum {
+    FP_SUCCESS = 0,
+    FP_FAILIURE = 1,
```
Comment (erlingrj) [inline programs/lib/include/flexpret_types.h:12]:
```suggestion
    FP_FAILURE = 1,
```

File: programs/lib/include/flexpret_noc.h [hunk 5]
```diff
@@ -13,26 +15,95 @@
...
+    }
+    timeout_t time = rdtime() + timeout;
+    while (rdtime() < time) {
+        if (NOC_TX_READY(NOC_CSR)) {
+            NOC_DEST = addr;
+            NOC_DATA = data;
+            return FP_SUCCESS;
+        }
+    }
+    return FP_FAILIURE;
```
Comment (erlingrj) [inline programs/lib/include/flexpret_noc.h:55]:
```suggestion
    return FP_FAILURE;
```

File: programs/lib/include/flexpret_noc.h [hunk 6]
```diff
@@ -13,26 +15,95 @@
...
+        NOC_DATA = data;
+        return FP_SUCCESS;
+    }
+    if (timeout == TIMEOUT_NEVER) {
+        if (NOC_TX_READY(NOC_CSR)) {
+            NOC_DEST = addr;
+            NOC_DATA = data;
+            return FP_SUCCESS;
+        } else {
+            return FP_FAILIURE;
```
Comment (erlingrj) [inline programs/lib/include/flexpret_noc.h:44]:
```suggestion
            return FP_FAILURE;
```

File: programs/lib/include/flexpret_noc.h [hunk 7]
```diff
@@ -13,26 +15,95 @@
...
+        while (!NOC_DATA_AVAILABLE(NOC_CSR));
+        *data = NOC_DATA;
+        return FP_SUCCESS;
+    }
+    if (timeout == TIMEOUT_NEVER) {
+        if (NOC_DATA_AVAILABLE(NOC_CSR)) {
+            *data = NOC_DATA;
+            return FP_SUCCESS;
+        }
+        return FP_FAILIURE;
```
Comment (erlingrj) [inline programs/lib/include/flexpret_noc.h:97]:
```suggestion
        return FP_FAILURE;
```

File: programs/lib/include/flexpret_noc.h [hunk 8]
```diff
@@ -13,26 +15,95 @@
...
+    }
+    uint32_t time = rdtime() + timeout;
+    while (rdtime() < time) {
+        if (NOC_DATA_AVAILABLE(NOC_CSR)) {
+            *data = NOC_DATA;
+            return FP_SUCCESS;
+        }
+    }
+    return FP_FAILIURE;
```
Comment (erlingrj) [inline programs/lib/include/flexpret_noc.h:106]:
```suggestion
    return FP_FAILURE;
```

**Review by erlingrj — APPROVED** [review]
Looks good to me (just suggested changes of a typo)

**Pull request merged**